hier_block2: handling processor affinity segfaults

dl1ksv opened the pull request:
<!--- The title of the PR should summarize the change implemented. -->
<!--- Example commit message format: -->
<!--- `module: summary of change` -->
<!--- (leave blank) -->
<!--- `details of what/why/how an issue was addressed` -->
<!--- Keep subject lines to 50 characters (but 72 is a hard limit!) -->
<!--- characters. Refer to the [Revision Control Guidelines](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md#revision-control-guidelines) section of the coding guidelines -->

## Description
<!--- Provide a general summary of your changes in the title above -->
<!--- Why is this change required? What problem does it solve? -->
In some cases set_processor_affinity / unset_processor_affinity / processor_affinity leads to a segfault. One example can be found in #6558.

The reason is an infinite loop in calling these functions.

This pr avoids this loop.
## Related Issue
<!--- Refer to any related issues here -->
<!--- If this PR fully addresses an issue, please say "Fixes #1234", -->
<!--- as this will allow Github to automatically close the related Issue -->
Fixes #6558
## Which blocks/areas does this affect?
<!--- Include blocks that are affected and some details on what -->
<!--- areas these changes affect, such as performance. -->
hier_block2 and hier_block2_detail
## Testing Done
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you -->
<!--- ran to see how your change affects other areas of the code, -->
<!--- etc. Then, include justifications for how your tests -->
<!--- demonstrate those affects. -->
Running the example in #6558
## Checklist
<!--- Go over all the following points, and put an `x` in all the
<!--- boxes that apply. Note that some of these may not be valid -->
<!--- for all PRs. -->

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [ ] I have added tests to cover my changes, and all previous tests pass.

willcode commented:
@dl1ksv @marcusmueller would you deconflict this with #6627?

marcusmueller commented:
That'll be hard! We're solving the same problem two different ways here!

marcusmueller commented on gnuradio-runtime/lib/hier_block2_detail.cc:
```diff
@@ -891,24 +891,24 @@ void hier_block2_detail::set_processor_affinity(const std::vector<int>& mask)
     for (basic_block_viter_t p = tmp.begin(); p != tmp.end(); p++) {
-        (*p)->unset_processor_affinity();
+        if (d_owner->unique_id() != (*p)->unique_id())
+            (*p)->unset_processor_affinity();
```
Won't this fail if this hier block contains hier block, whose message output connects to an input of this hier block?

dl1ksv replied:
I tested this with the example you added to qa_hier_block2.py and it does not seqfault.

marcusmueller commented on gnuradio-runtime/include/gnuradio/hier_block2.h:
```diff
@@ -45,6 +45,7 @@ class GR_RUNTIME_API hier_block2 : public basic_block
      */
     std::unique_ptr<hier_block2_detail> d_detail;
 
+    std::vector<int> d_affinity; // thread affinity proc. mask
```
why keep this in the interface instead of in the detail, if it's private anyways?

willcode replied:
I was thinking the same thing ... but there are worse offenders in this file.

dl1ksv replied:
It's the same behaviour as in block.k. See line 728 and 799

dl1ksv replied:
And you need it to save and clear the mask if you call the affinity functions for the hier block.

willcode commented:
@dl1ksv @marcusmueller Discussions in both threads ended in a non-reassuring way :smile: . Any progress on either of these, or any guidance?

dl1ksv commented:
I now fixed the two segfaults in my proposed solution and added the test additional test from #6627 and completed it with calls for processor_affinity() and log_level().

Both calls fail at the moment #6627 but I think this could easily been fixed.
So I have no idea which solution should be prefered.

dl1ksv commented:
The qa_logger test seems to be affected by settings in ~/.gnuradio/config.conf. I have to figure that out.

dl1ksv commented:
The log_level handling of the top_block is buggy in main. Just building a top_block and printing or setting the log level results in seqfault

tb=top_block() 
print(tb.log_level())  # Seqfault !!!
tb.set_log_level('debug') # Seqfault !!!

I would expect at this point either the configuration setting of 
  $CMAKE_INSTALL_PREFIX/etc/gnuradio/conf.d/gnuradio-runtime.conf or
~/.gnuradio/config.conf if exists is used.

But you have to connect blocks to the top block, to make these calls work.

Why does only changing the log level of the first build block affects the log level of the top block and no others?
And it can quickly become confusing , who is the first.
Changing the log level of the top block by set_log_level affects all depend blocks, but not if the it's level is changed by setting the level of the first depending block.

I propose to restrict set_log_level to the block it's applied to and it's depend blocks. So I modified the qa_logger.py  to follow these thoughts.
